### MongoDB 

In [ ]:
import vega_datasets
dir(vega_datasets.data)

['7zip',
 'airports',
 'annual_precip',
 'anscombe',
 'barley',
 'birdstrikes',
 'budget',
 'budgets',
 'burtin',
 'cars',
 'climate',
 'co2_concentration',
 'countries',
 'crimea',
 'disasters',
 'driving',
 'earthquakes',
 'ffox',
 'flare',
 'flare_dependencies',
 'flights_10k',
 'flights_200k',
 'flights_20k',
 'flights_2k',
 'flights_3m',
 'flights_5k',
 'flights_airport',
 'gapminder',
 'gapminder_health_income',
 'gimp',
 'github',
 'graticule',
 'income',
 'iowa_electricity',
 'iris',
 'jobs',
 'la_riots',
 'londonBoroughs',
 'londonCentroids',
 'londonTubeLines',
 'lookup_groups',
 'lookup_people',
 'miserables',
 'monarchs',
 'movies',
 'normal_2d',
 'obesity',
 'ohlc',
 'points',
 'population',
 'population_engineers_hurricanes',
 'seattle_temps',
 'seattle_weather',
 'sf_temps',
 'sp500',
 'stocks',
 'udistrict',
 'unemployment',
 'unemployment_across_industries',
 'uniform_2d',
 'us_10m',
 'us_employment',
 'us_state_capitals',
 'volcano',
 'weather',
 'weball26',
 'wheat',

In [ ]:
# Selecting 3 datasets to test

# 1. windvectors(4800 rows × 5 columns)
# 2. volcano (5307 rows × 3 columns)
# 3. sf_temps (8759 rows × 2 columns)


### Loading data to Mongodb by single dataset

In [1]:
# 1.windvectors dataset

import vega_datasets
import pymongo
import time

# remove ****** and fill your mongodb password in connString
connString='mongodb://sathesht11:*******@ac-pdkcaeq-shard-00-00.'\
            'p3oqjnt.mongodb.net:27017,ac-pdkcaeq-shard-00-01.p3oqjnt.'\
            'mongodb.net:27017,ac-pdkcaeq-shard-00-02.p3oqjnt.'\
            'mongodb.net:27017/?ssl=true&replicaSet=atlas-jw3sl9-shard-'\
            '0&authSource=admin&retryWrites=true&w=majority'

def change_df_to_dict(df):
  data = df.to_dict('records')
  return data

def load_data_to_mongodb(df,connString,col_name):
  data = change_df_to_dict(df)
  conn = pymongo.MongoClient(connString)
  db = conn['vega_datasets']
  col = db[col_name]
  for i in data:
    col.insert_one(i)

start_time=time.time()
load_data_to_mongodb(vega_datasets.data.windvectors(),connString,'windvectors')
end_time=time.time()
total_time=round((end_time-start_time)/60,3)
print('windvectors_data completed')
print('Time taken:',total_time,'mins')

windvectors_data completed
Time taken: 13.116645602385203 mins


In [5]:
# 2.volcano dataset
import vega_datasets
import pymongo
import time

# remove ****** and fill your mongodb password in connString
connString='mongodb://sathesht11:*******@ac-pdkcaeq-shard-00-00.'\
            'p3oqjnt.mongodb.net:27017,ac-pdkcaeq-shard-00-01.p3oqjnt.'\
            'mongodb.net:27017,ac-pdkcaeq-shard-00-02.p3oqjnt.'\
            'mongodb.net:27017/?ssl=true&replicaSet=atlas-jw3sl9-shard-'\
            '0&authSource=admin&retryWrites=true&w=majority'

def change_df_to_dict(df):
  data = df.to_dict('records')
  return data

def load_data_to_mongodb(df,connString,col_name):
  data = change_df_to_dict(df)
  conn = pymongo.MongoClient(connString)
  db = conn['vega_datasets']
  col = db[col_name]
  for i in data:
    col.insert_one(i)

start_time=time.time()
load_data_to_mongodb(vega_datasets.data.volcano(),connString,'volcano')
end_time=time.time()
total_time=round((end_time-start_time)/60,3)
print('volcano_data completed')
print('Time taken:',total_time,'mins')

volcano_data completed
Time taken: 14.498 mins


In [4]:
# 3.sf_temps dataset
import vega_datasets
import pymongo
import time

# remove ****** and fill your mongodb password in connString
connString='mongodb://sathesht11:*******@ac-pdkcaeq-shard-00-00.'\
            'p3oqjnt.mongodb.net:27017,ac-pdkcaeq-shard-00-01.p3oqjnt.'\
            'mongodb.net:27017,ac-pdkcaeq-shard-00-02.p3oqjnt.'\
            'mongodb.net:27017/?ssl=true&replicaSet=atlas-jw3sl9-shard-'\
            '0&authSource=admin&retryWrites=true&w=majority'

def change_df_to_dict(df):
  data = df.to_dict('records')
  return data

def load_data_to_mongodb(df,connString,col_name):
  data = change_df_to_dict(df)
  conn = pymongo.MongoClient(connString)
  db = conn['vega_datasets']
  col = db[col_name]
  for i in data:
    col.insert_one(i)

start_time=time.time()
load_data_to_mongodb(vega_datasets.data.sf_temps(),connString,'sf_temps')
end_time=time.time()
total_time=round((end_time-start_time)/60,3)
print('sf_temps_data completed')
print('Time taken:',total_time,'mins')

sf_temps_data completed
Time taken: 23.895 mins


## Loading multiple datasets to MongoDB using Multi-Threading

In [6]:
import vega_datasets
import pymongo

# remove ****** and fill your mongodb password in connString
connString='mongodb://sathesht11:*******@ac-pdkcaeq-shard-00-00.'\
            'p3oqjnt.mongodb.net:27017,ac-pdkcaeq-shard-00-01.p3oqjnt.'\
            'mongodb.net:27017,ac-pdkcaeq-shard-00-02.p3oqjnt.'\
            'mongodb.net:27017/?ssl=true&replicaSet=atlas-jw3sl9-shard-'\
            '0&authSource=admin&retryWrites=true&w=majority'

def change_df_to_dict(df):
  data = df.to_dict('records')
  return data

def load_data_to_mongodb(df,connString,col_name):
  data = change_df_to_dict(df)
  conn = pymongo.MongoClient(connString)
  db = conn['vega_datasets']
  col = db[col_name]
  for i in data:
    col.insert_one(i)

def windvectors_data():
  load_data_to_mongodb(vega_datasets.data.windvectors(),connString,'windvectors')
  print('windvectors_data completed')

def volcano_data():
  load_data_to_mongodb(vega_datasets.data.volcano(),connString,'volcano')
  print('volcano_data completed')

def sf_temps_data():
  load_data_to_mongodb(vega_datasets.data.sf_temps(),connString,'sf_temps')
  print('sf_temps_data completed')


import threading
from threading import *
import time

start_time = time.time()

x1=Thread(target=windvectors_data)
x1.start()

x2=Thread(target=volcano_data)
x2.start()

x3=Thread(target=sf_temps_data)
x3.start()


x1.join()
x2.join()
x3.join()


end_time = time.time()
total_time=round((end_time-start_time)/60,3)

print('Time taken:',total_time,'mins')


windvectors_data completed
volcano_data completed
sf_temps_data completed
Time taken: 23.917 mins


    windvectors dataset loaded time = 13.116 mins

        volcano dataset loaded time = 14.498 mins
    
       sf_temps dataset loaded time = 23.895 mins
   

    So processing time of all datasets: 51.509 mins

     But Multi-Threading method's time: 23.917 mins 
 
     As a result, the Multi-Threading method demonstrates the effectiveness of its usage.